hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -442,6 +450,93 @@ def runTest(self):
         StateMachineTestCase.__qualname__ = cls.__qualname__ + ".TestCase"
         return StateMachineTestCase
 
+    def _execute_fn(self, rule, stateful_run_times, **data):
+        result = get_sync_result_with_time(
+            rule.function, stateful_run_times, self, **data
+        )
+        return result
+
+
+class TrioRuleBasedStateMachine(RuleBasedStateMachine):
+    def __init__(self):
+        super().__init__()
+
+    def _execute_fn(self, rule, stateful_run_times, **data):
+        if inspect.iscoroutinefunction(rule.function):
+            assert hasattr(self, "nursery") and isinstance(self.nursery, trio.Nursery)
+            formatted_task = partial(
+                get_async_result_with_time,
+                rule.function,
+                stateful_run_times,
+                self,
+                **data,
+            )
+            self.nursery.start_soon(formatted_task)
+
+            # TODO: At the moment, we can't return any values for async functions. Should this remain the case?
+            # Trio doesn't seem to (https://stackoverflow.com/questions/51567451/capture-the-return-value-from-nursery-objects).
+            # Maybe include an error for `return` keywords in async @rules
+            return
+
+        return super()._execute_fn(rule, stateful_run_times, **data)
+
+    def loop_setup(self, fn):
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+            async def run_with_nursery(*args, **kwargs):
+                async with trio.open_nursery() as nursery:
+                    self.nursery = nursery
+                    fn(*args, **kwargs)
+
+            return trio.run(run_with_nursery, *args, **kwargs)
+
+        return wrapper
+
+def async_manager_decorator(fn):
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        machine = args[0]
+        if hasattr(machine, "loop_setup") and callable(machine.loop_setup):
+            return machine.loop_setup(fn)(*args, **kwargs)
+        return fn(*args, **kwargs)
+
+    return wrapper
+
+
+def get_sync_result_with_time(func, stateful_run_times, *args, **kwargs):
+    label = f"execute:rule:{func.__name__}"
+    start = perf_counter()
+    start_gc = gc_cumulative_time()
+
+    result = func(*args, **kwargs)
+
+    in_gctime = gc_cumulative_time() - start_gc
+    duration = perf_counter() - start - in_gctime
+    stateful_run_times[label] = stateful_run_times.get(label, 0) + duration
+    print(
+        f"{label} (sync) took {duration:.4f} seconds (excluding GC: {in_gctime:.4f} seconds)"
+    )
+
+    return result
+
+
+async def get_async_result_with_time(func, stateful_run_times, *args, **kwargs):
+    """This is mainly a copy of the sync version of the method, but plays nicely with python async/await"""
```

Comment:
I think it would be possible to move most of the duplicated logic into a (sync-only) context manager
```
with recorded_time(stateful_run_times, label=f"execute:rule:{func.__name__}"):
    return await func(*args, **kwargs)
```

Reply:
Oh that's much cleaner, thanks @jobh!